### ISSUE TITLE
Bulk Enrollment for Courses & Memberships

### ISSUE THREAD

**Issue by actual-saurabh** [gocodebox/lifterlms#562]
- [x] Using `manage_users_extra_tablenav` action here: https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/class-wp-users-list-table.php#L300, it'd be easy to add a `select2` box for memberships and courses and have a simple UI for bulk enrollment. The choice would be limited to only one course or membership at a time.

- [x] Set a custom administrative *action* called `'enroll'` (https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/class-wp-list-table.php#L478-L499 & https://github.com/WordPress/WordPress/blob/master/wp-admin/users.php#L99

- [x] Filter the `default` case of the switch statement here: https://github.com/WordPress/WordPress/blob/master/wp-admin/users.php#L444 to get the array of `user_ids` and the rest of the `$_POST` data and then loop through each user and run `llms_enroll_student`https://github.com/gocodebox/lifterlms/blob/master/includes/functions/llms.functions.person.php#L156-L171 to complete enrollment

**Comment by actual-saurabh** [comment]
Looks like the last 2 are not possible in the way I assumed they would be. However, I was able to find workarounds and am still able to run the bulk enrollment.

Will be able to push this in a day or so.

**Comment by actual-saurabh** [comment]
Finished all the code for bulk enrollment here: https://github.com/actual-saurabh/lifterlms/blob/feature/bulk-enroll/includes/admin/class.llms.student.bulk.enroll.php

Added select2 extension here: https://github.com/actual-saurabh/lifterlms/blob/feature/bulk-enroll/assets/js/llms-bulk-enroll.js

Wasn't able to test thoroughly but the js doesn't seem to load at all. Will test and fine tune before sending in a PR

**Comment by actual-saurabh** [comment]
Tested all scenarios manually by adding the js through console. Works beautifully 😄 :

Select users and memberships:
![screen shot 2018-06-14 at 6 31 07 pm](https://user-images.githubusercontent.com/1739834/41413626-45dd05e0-7001-11e8-9c0a-bf8ce52dca16.png)

Everybody gets enrolled:
![screen shot 2018-06-14 at 6 31 24 pm](https://user-images.githubusercontent.com/1739834/41413652-59725182-7001-11e8-8b7e-7d30b420b290.png)

Select membership, but not users:
![screen shot 2018-06-14 at 6 33 25 pm](https://user-images.githubusercontent.com/1739834/41413746-9df1db2a-7001-11e8-84a0-4daeafdb6d5d.png)

Expected error!
![screen shot 2018-06-14 at 6 33 39 pm](https://user-images.githubusercontent.com/1739834/41413753-a3ef6d8a-7001-11e8-9ef6-941497a0e58a.png)

Select users but no product
![screen shot 2018-06-14 at 6 34 11 pm](https://user-images.githubusercontent.com/1739834/41413762-aaedd45a-7001-11e8-951e-6eba5d6e7280.png)

Expected error!
![screen shot 2018-06-14 at 6 34 22 pm](https://user-images.githubusercontent.com/1739834/41413768-aed2bcb6-7001-11e8-9f54-4ab1e3d338bb.png)

Try enrolling already enrolled students
![screen shot 2018-06-14 at 6 37 28 pm](https://user-images.githubusercontent.com/1739834/41413885-1bec4fe2-7002-11e8-8289-35ff394b3347.png)

Expected error!
![screen shot 2018-06-14 at 6 37 45 pm](https://user-images.githubusercontent.com/1739834/41413893-2396bc1e-7002-11e8-90ce-a34b6fea1831.png)

**Comment by actual-saurabh** [comment]
A side thought, similar functionality can be added to the Edit User screen to bulk enroll one user into many courses/memberships with a few tweaks.

**Comment by thomasplevy** [comment]
@actual-saurabh I'm wondering if it would be better to group the notices into a single notice with a `<ul>` instead of multiple notices. 3 - 6 notices isn't that bad but what if a user enrolls 20? What if they extend the default number of users / screen to 50 or 100 and then bulk enroll all 50 / 100? That's a lot of notices... I guess it'd be a giant list as well but at least it could be dismissed in one click instead of 20 - 100.

Just a thought

**Comment by actual-saurabh** [comment]
I did think about this. I considered 3 scenarios:

1. Frequent bulk enrollment for a small number of users (~10)
2. Rare bulk enrollment for a large number of students (>10, <20-30 )
3. Extremely rare bulk enrollment for a humongous number (>50)

Most common use cases, afaik, would be 1 & 2. If someone does perform 3, that'll be a one off, very rare user experience.

I felt that since each enrollment is a separate operation, it makes sense to keep the notice separate to distinguish the state (success, error) in line with WordPress's default functionality. I'm banking on WordPress core taking care of grouping notifications in the near future. See: https://core.trac.wordpress.org/ticket/43484

In any case, simply refreshing the page or navigating away from the screen would clear all the notifications by default.

So, I decided to keep it as it is, even if it is unelegant, for the sake of consistency with WP's admin experience. In that vein,  I feel it is more consistent with the idea of being able to clear the success notices and have only the failed notifications on screen, if and when enrollment fails for some users.
